src/tools/rbd_mirror/InstanceWatcher.cc

```diff
@@ -349,6 +445,266 @@ void InstanceWatcher<I>::notify_image_release(
 }
 
 template <typename I>
+void InstanceWatcher<I>::notify_start_sync(const std::string &sync_id,
+                                           Context *on_notify_ack) {
+  dout(20) << "sync_id=" << sync_id << dendl;
+
+  Mutex::Locker locker(m_lock);
+
+  if (m_instance_sync_throttler != nullptr) {
+    handle_sync_start(m_instance_id, sync_id, on_notify_ack);
+  } else {
+    send_sync_start(sync_id, on_notify_ack);
+  }
+}
+
+template <typename I>
+void InstanceWatcher<I>::send_sync_start(const std::string &sync_id,
+                                         Context *on_start) {
+  dout(20) << "sync_id=" << sync_id << dendl;
+
+  assert(m_lock.is_locked());
+
+  if (m_remote_throttler_syncs.count(sync_id) == 0) {
+    uint64_t request_id = ++m_request_seq;
+
+    bufferlist bl;
+    ::encode(NotifyMessage{SyncStartPayload{request_id, sync_id}}, bl);
+
+    auto sync_ctx = new C_SyncRequest(this, sync_id, on_start);
+    sync_ctx->req = new C_NotifyInstanceRequest(this, "", request_id,
+                                                std::move(bl), sync_ctx);
+
+    m_remote_throttler_syncs[sync_id] = sync_ctx;
+    sync_ctx->req->send();
+    return;
+  }
+
+  // This should only be possible when notify_finish_sync for the
```

Comment:
A user completes a sync by sending `notify_finish_sync` (which returns immediately, after queuing the sync notification to the leader), and after this (by some reason) calls `notify_start_sync` for the same image again. At this moment the previous notification for this sync_id may still be in progress.

This could be avoided by using some unique value for sync_id instead of image_id by the caller, but this will complicate things on the caller side.